larq/quantizers.py

```diff
@@ -450,6 +452,90 @@ def __init__(
         )
 
 
+@utils.register_keras_custom_object
+class ReluSign(QuantizerFunctionWrapper):
+    r"""
+    Instantiates a binarization quantizer with output values 0 and 1.
+    \\[
+    q(x) = \begin{cases}
+    +1 & x > \Delta \\\
+    0 & x \leq \Delta
+    \end{cases}
+    \\]
+
+    where $\Delta$ is defined as the threshold and can be passed as an argument.
+    The gradient is estimated using the Straight-Through Estimator
+    (essentially the binarization is replaced by a clipped identity on the
+    backward pass).
+
+    \\[\frac{\partial q(x)}{\partial x} = \begin{cases}
+    1 & \left|x\right| \leq 1 \\\
+    0 & \left|x\right| > 1
+    \end{cases}\\]
+
+    ```plot-activation
+    quantizers.relu_sign
+    ```
+
+    # Arguments
+    threshold_value: The value for the threshold, $\Delta$, default value 0.
+    clip_value: Threshold for clipping gradients.
+
+    # Returns
+    AND Binarization function
+    """
+
+    def __init__(self, threshold_value=0.0, clip_value=1.0):
+        super().__init__(relu_sign, threshold_value=threshold_value, clip_value=clip_value)
+
+
+@utils.register_keras_custom_object
+@utils.set_precision(1)
+def relu_sign(x, threshold_value=0.0, clip_value=1.0):
+    r"""
+    Binarization function with output values 0 and 1.
+
+    \\[
+    q(x) = \begin{cases}
+    +1 & x > \Delta \\\
+    0 & x \leq \Delta
+    \end{cases}
+    \\]
+
+    where $\Delta$ is defined as the threshold and can be passed as an argument.
+    The gradient is estimated using the Straight-Through Estimator
+    (essentially the binarization is replaced by a clipped identity on the
+    backward pass).
+
+    \\[\frac{\partial q(x)}{\partial x} = \begin{cases}
+    1 & \left|x\right| \leq 1 \\\
+    0 & \left|x\right| > 1
+    \end{cases}\\]
+
+    ```plot-activation
+    quantizers.relu_sign
+    ```
+
+    # Arguments
+    x: Input tensor.
+    threshold_value: The value for the threshold, $\Delta$, default value 0.
+    clip_value: Threshold for clipping gradients.
+
+    # Returns
+    AND-binarized tensor.
+    """
+    x = tf.clip_by_value(x, -clip_value, clip_value)
+
+    @tf.custom_gradient
+    def _and_binarize_with_identity_grad(x):
+        def grad(dy):
```

Comment:
This is not something I found in the literature, it is something that I put together myself. Do we need a reference?

Reply:
No, we don't need a reference. I was just curious, since I'd expect the pseudo-gradient to be more an approximation of the sigmoid gradient rather then an approximation of the hard_tanh gradient.

Reply:
I did not experiment with having a different gradient approximation. There are different possibilities for approximating the gradient. I picked this one because it is the easiest and the clipped equivalent of the ste of the hard_tanh would be something like a scaled version of this with a wider clip interval. So it does not seem like there would be a big difference. Anyways I think we want to spend some time on looking closer into AND quantization and its gradient. We can put this PR on hold in case you think this needs to be done first.